File: algorithms/java/ShellSort.java
package algorithms.java;

public class ShellSort {
    public static void sort(int[] arr) {
        int n = arr.length;

        int gap = 1;
        while (gap < n / 3) {
            gap = 3 * gap + 1;
        }

        while (gap > 0) {
            for (int i = gap; i < n; i++) {
                int temp = arr[i];
                int j = i;
                while (j >= gap && arr[j - gap] > temp) {
                    arr[j] = arr[j - gap];
                    j -= gap;
                }
                arr[j] = temp;
            }
            gap /= 3;
        }
    }
}

File: algorithms/go/shellSort.go
package sortgo

func ShellSort(arr []int) {
	n := len(arr)

	gap := 1
	for gap < n/3 {
		gap = 3*gap + 1
	}

	for gap > 0 {
		for i := gap; i < n; i++ {
			temp := arr[i]
			j := i
			for j >= gap && arr[j-gap] > temp {
				arr[j] = arr[j-gap]
				j -= gap
			}
			arr[j] = temp
		}
		gap /= 3
	}
}
